### DISCUSSION TITLE
Did you try SGDK_wine?

### DISCUSSION THREAD

**Discussion by Franticware** [Franticware/SGDK_wine#4] [General]
If you did, please share the following info in the comments:
1. Does the script work for you?
2. Your Linux distribution - which one, version, variant (32/64-bit), ... 
2. Script revision (line 3 of the script)
3. SGDK version

Thank you!

**Comment by gwleckyt** [comment]
Amazing. Works on a Chromebook! Model CB3-431 (Acer Chromebook 14)

1. Works fine. Compiled helloWorld from the SGDK 1.60 tutorial, resulting ROM works in BlastEm (Linux).
2. uname -r returns 5.4.67, uname -m returns x86_64, os-release reports Debian GNU/Linux 10 (buster)
3. revision 1.60-2
4. 1.60

Note: I have Java runtimes installed from previous experiments building the SGDK toolchain from scratch (failed), testing against SGDK 1.51 downloaded as part of the gendev runtimes (okay, but wanted 1.60).

Good job, by the way, I'm a happy coder.

**Comment by pimstolk** [comment]
1  Script works perfectly! 
2 Ubuntu 20.04 5.8.0-44
3 1.60-2
4 1.60


Thanks!

**Comment by Rafagars** [comment]
1. Works fine 
2. Linux Mint 20.1
3. 1.60-2
4. 1.60 (Didn't work for 1.62)

**Reply by Franticware** [reply]
Thank you for the report, I have just updated the script. According to preliminary testing, it now works with 1.62 too. I will test it more.

**Reply by Rafagars** [reply]
Works fine. Thanks!

**Comment by ReaderGuy42** [comment]
Hi, I'm on Linux Mint 20.2 and I'm having issues. I should mention that I'm still fairly new to Linux, so keep that in mind, but on your second line `cp generate_wine.sh $PATH_TO_SGDK/bin` I get the error that it `cp: cannot stat 'generate_wine.sh': No such file or directory`
My "path" looks like this `PATH_TO_SGDK=~/Documents/SGDK-1.62` - Is that correct?
Any pointers for a beginner?
Thanks!

**Reply by Franticware** [reply]
Do you have generate_wine.sh present in your working directory? The terminal opens in the home directory by default, so first download generate_wine.sh to your home directory, then try the steps again.

**Reply by ReaderGuy42** [reply]
I'm not sure I understand. I have SGDK setup in /opt/SGDK-1.62 and put generate_wine.sh in the bin folder within the SGDK-1.62 folder. When I then do `cp generate_wine.sh $PATH_TO_SGDK/bin` I get the output 
`cp generate_wine.sh $PATH_TO_SGDK/bin
cp: 'generate_wine.sh' and '/opt/SGDK-1.62/bin/generate_wine.sh' **are the same file**`

**Reply by Hockmiester** [reply]
Similar problem. I tried on Mint/Raspberry Pi/Twister. 
- I'm using a rpi4 with twisters> I'm flexible on OS if I've installed the wrong distro 
- I installed JDK-8u321-linux-aarch64 1st 
- I downloaded the generate_wine.sh from the main github page (I assume that was correct)
- Placed it in the bin folder
- Navigated to ~/sdk170/bin (where the .sh is)
- Ran sh generate_wine.sh
-  Got this error: generate_wine.sh: 8: generate_wine.sh: Syntax error: newline unexpected

Also tried & failed: installing Java 1.21 (1999 release) from Wine Desktop. Trying to run SGDK from Wine Desktop using the cmd. 

I'm not a developer, I'm playing around for fun because I have excess time on my hands and trying to learn things in the void.

**Reply by Franticware** [reply]
You might want to try recently updated script. The one you tried was not compatible with SGDK 1.70.

**Comment by nicomegadrive** [comment]
I'm using it every day and it works well so far, I'm on Ubuntu 22.10 (64 bits), I have the latest version of sgdk_wine and the latest version of SGDK(1.8).

Thanks for this script, it's really helpful, I didn't see myself running a VM just for SGDK.

One thing I noticed is that the script won't see the changes I made in the sprites / tiles (width / height / colors for example), if I don't edit the resources.res first.

**Comment by thomasgoldstein** [comment]
Hi folks. I am using Linux Mint (Linux Mint 21.1 Cinnamon). I'm very much a Linux noob, but I ended up getting it to work after a bit of time. I'm giving all possible details in case this could lead to the instructions getting improved.

First, I thought this script was supposed to be "generate_wine.sh":

```
PATH_TO_SGDK=/path/to/sgdk
cp generate_wine.sh $PATH_TO_SGDK/bin
cd $PATH_TO_SGDK/bin
sh generate_wine.sh
```

I was confused because that script is not named explicitly, and I wasn't even sure if it needed to be in a file or if it could be run from the command line directly. Then I realized the generate_wine.sh file is the one in the Git repo (duh). I downloaded it, put it in my /sgdk/bin, then created a new file called "generate_wine_start.sh" containing the 4 lines above, in the same folder. I ran generate_wine_start.sh, and it seems it did its job fine as it ended up outputting this (the files it created):

> ar
> as
> bintos
> cc1
> cp
> cpp
> gcc
> gdb
> ld
> lto1
> lto-wrapper
> mac68k
> make
> mkdir
> nm2wch
> nm
> objcopy
> objdump
> rm
> sh
> size
> sjasm
> xgmtool

So far so good, but then I reach this command line:

`make GDK=/path/to/sgdk -f /path/to/sgdk/makefile_wine.gen`

And I'm confused again, because the readme says "The following command should then be able to build any SGDK project:", while I don't see a path to an SGDK project in that command line. How am I supposed to let it know where my SGDK project is? It only mentions the path to SGDK itself.

I still tried to run it to see if I could get a response that'd lead me to the solution, and I got this:

> mkdir -p src/boot
> mkdir -p out
> cp /home/blabla/sgdk/src/boot/sega.s src/boot/sega.s
> cp /home/blabla/sgdk/src/boot/rom_head.c src/boot/rom_head.c
> /home/blabla/sgdk/bin/gcc  -DSGDK_GCC -m68000 -Wall -Wextra -Wno-shift-negative-value -Wno-main -Wno-unused-parameter -fno-builtin -fms-extensions -Iinc -Isrc -Ires -I/home/blabla/sgdk/inc -I/home/blabla/sgdk/res -B/home/blabla/sgdk/bin -c src/boot/rom_head.c -o out/rom_head.o
> /home/blabla/sgdk/bin/gcc: line 3: wine: command not found
> make: *** [/home/blabla/sgdk/makefile_wine.gen:174: out/rom_head.o] Error 127

Oh, right, I did not have Wine installed. I decided not to install it early in case it'd get done automatically by the script or something. I installed Wine myself, then tried again. Still no good:

> mkdir -p src/boot
> mkdir -p out
> /home/blabla/sgdk/bin/gcc  -DSGDK_GCC -m68000 -Wall -Wextra -Wno-shift-negative-value -Wno-main -Wno-unused-parameter -fno-builtin -fms-extensions -Iinc -Isrc -Ires -I/home/blabla/sgdk/inc -I/home/blabla/sgdk/res -B/home/blabla/sgdk/bin -c src/boot/rom_head.c -o out/rom_head.o
> wine: created the configuration directory '/home/blabla/sgdk/bin/.wineconf'
> MESA-INTEL: warning: Haswell Vulkan support is incomplete
> Could not find Wine Gecko. HTML rendering will be disabled.
> Could not find Wine Gecko. HTML rendering will be disabled.
> wine: configuration in L"/home/blabla/sgdk/bin/.wineconf" has been updated.
> /home/blabla/sgdk/bin/objcopy -O binary out/rom_head.o out/rom_head.bin
> /home/blabla/sgdk/bin/gcc -x assembler-with-cpp -Wa,--register-prefix-optional,--bitwise-or  -DSGDK_GCC -m68000 -Wall -Wextra -Wno-shift-negative-value -Wno-main -Wno-unused-parameter -fno-builtin -fms-extensions -Iinc -Isrc -Ires -I/home/blabla/sgdk/inc -I/home/blabla/sgdk/res -B/home/blabla/sgdk/bin -O3 -fuse-linker-plugin -fno-web -fno-gcse -fno-unit-at-a-time -fomit-frame-pointer -flto -c src/boot/sega.s -o out/sega.o
> echo "" > out/cmd_
> /home/blabla/sgdk/bin/gcc -B/home/blabla/sgdk/bin -n -T /home/blabla/sgdk/md.ld -nostdlib out/sega.o @out/cmd_ /home/blabla/sgdk/lib/libmd.a /home/blabla/sgdk/lib/libgcc.a -o out/rom.out -Wl,--gc-sections
> C:\users\blabla\Temp\cclgCHHF.ltrans1.ltrans.o: In function `_reset_entry':
> <artificial>:(.text+0x2a0): undefined reference to `main'
> C:\users\blabla\Temp\cclgCHHF.ltrans1.ltrans.o: In function `_start_entry':
> <artificial>:(.text+0x1378): undefined reference to `main'
> <artificial>:(.text+0x144c): undefined reference to `main'
> make: *** [/home/blabla/sgdk/makefile_wine.gen:161: out/rom.out] Error 1

I guessed my problem was related to the fact I had not specified the path to my SGDK project, so it could not resolve that "out" folder. But I was still not sure what to do about it. I had the following ideas:
- Add the SGDK project path as a command line parameter? I guess not, or it'd have been in the readme where that line is.
- Modify the makefile_wine.gen file to specify the path to my SGDK project? I had a look at the file content, and although it may work, that seems error prone as the "out" folder is referenced in multiple places.
- Maybe I'm supposed to copy that makefile_wine.gen file in my SGDK project folder and call it there?

That last idea did the trick! I'm not even sure if that's how it's supposed to be done, but it worked.

After fixing some paths in my project resources (replacing some \ with / in wav or png paths), now I can compile my project and it generates a ROM file.

Last struggle, but which is out of the scope of SGDK_wine: when trying to load the generated ROM in BlastEm, it would show a black screen and freeze. I figured this is because of some permission issue, BlastEm not being able to access the ROM file, so I guess I'm supposed to change the permission to the ROM file and/or parent folder. When I copied that ROM to Windows, I could see that it works fine.

Hope that helps, and thanks for your work.

Edit: I don't really know why BlastEm would not work, but the nightly works fine here! ([blastem64-0.6.3-pre-215c2afbe896.tar.gz](https://www.retrodev.com/blastem/nightlies/blastem64-0.6.3-pre-215c2afbe896.tar.gz))

**Comment by FBgamerisk** [comment]
1. Working perfectly
2. Debian 12
3. 1.80-1
4. SGDK 2.00
